filament/backend/src/vulkan/VulkanImageUtility.cpp

```diff
@@ -0,0 +1,245 @@
+/*
+ * Copyright (C) 2023 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "VulkanImageUtility.h"
+
+#include "VulkanTexture.h"
+
+#include <utils/Panic.h>
+#include <utils/algorithm.h>
+#include <utils/debug.h>
+
+#include <tuple>
+
+using namespace bluevk;
+
+namespace filament::backend {
+
+namespace {
+
+inline VkImageLayout getVkImageLayout(VulkanLayout layout) {
+    switch (layout) {
+        case VulkanLayout::UNDEFINED:
+            return VK_IMAGE_LAYOUT_UNDEFINED;
+        case VulkanLayout::READ_WRITE:
+            return VK_IMAGE_LAYOUT_GENERAL;
+        case VulkanLayout::READ_ONLY:
+            return VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL;
+        case VulkanLayout::TRANSFER_SRC:
+            return VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL;
+        case VulkanLayout::TRANSFER_DST:
+            return VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL;
+        case VulkanLayout::DEPTH_ATTACHMENT:
+            return VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL;
+        case VulkanLayout::DEPTH_SAMPLER:
+            return VK_IMAGE_LAYOUT_GENERAL;
+        case VulkanLayout::PRESENT:
+            return VK_IMAGE_LAYOUT_PRESENT_SRC_KHR;
+        // Filament sometimes samples from one miplevel while writing to another level in the same
+        // texture (e.g. bloom does this). Moreover we'd like to avoid lots of expensive layout
+        // transitions. So, keep it simple and use GENERAL for all color-attachable textures.
+        case VulkanLayout::COLOR_ATTACHMENT:
```

Comment:
could we do this only when a texture is both an attachment and sampleable?

Reply:
I tried to keep close to the original implementation for this first pass so that no extra errors are introduced.  But this is a good point and I've put a TODO to follow-up on separating the two cases.